File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.cpp
class Solution {
public:
    long long maxTotal(vector<int>& nums, string s) {
         int n=nums.size();
        long long pr=0;
        long long prs=-1000000000000000000LL;
        for(int i=0;i<n;i++)
        {
            bool has=(s[i]=='1');
            long long ne,sg;
            if(!has)
            {
                ne=pr>prs?pr:prs;
                sg=-1000000000000000000LL;
            }
            else
            {
                sg=(pr>prs?pr:prs)+nums[i];
                if(i>0)
                    ne=pr+nums[i-1];
                else
                    ne=-1000000000000000000LL;
            }
            pr=ne;
            prs=sg;
        }
        return pr>prs?pr:prs;
    }
};

File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.java
class Solution {
    public long maxTotal(int[] nums, String s) {
        int n = nums.length;

        long neg = Long.MIN_VALUE / 4;
        long[] dp = {0, neg};

        for (int i = 0; i < n; i++) {
            long[] ndp = {neg, neg};

            if (s.charAt(i) == '1') {
                for (int p = 0; p < 2; p++) {
                    if (dp[p] == neg) continue;

                    long add = (i > 0) ? nums[i - 1] : 0;
                    ndp[0] = Math.max(ndp[0], dp[p] + add);

                    add = (i > 0 && p == 1) ? nums[i - 1] : 0;
                    ndp[1] = Math.max(ndp[1], dp[p] + add);
                }
            } else {
                for (int p = 0; p < 2; p++) {
                    if (dp[p] == neg) continue;

                    long add = (i > 0 && p == 1) ? nums[i - 1] : 0;
                    ndp[0] = Math.max(ndp[0], dp[p] + add);
                }
            }

            dp = ndp;
        }

        return Math.max(dp[0], dp[1] + nums[n - 1]);
    }
}